Source/Relativity.Test.Helpers/Relativity.Test.Helpers/Kepler/OAuth2Helper.cs

```diff
@@ -0,0 +1,164 @@
+﻿using kCura.Relativity.Client;
+using Newtonsoft.Json.Linq;
+using Relativity.Services.Security;
+using Relativity.Services.Security.Models;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading.Tasks;
+using DTOs = kCura.Relativity.Client.DTOs;
+
+namespace Relativity.Test.Helpers.Kepler
+{
+	public class OAuth2Helper : IOAuth2Helper
+	{
+		private readonly IOAuth2ClientManager _oAuth2ClientManager;
+		private readonly IRSAPIClient _rsapiClient;
+
+		public OAuth2Helper(IOAuth2ClientManager oAuth2ClientManager, IRSAPIClient rsapiClient)
+		{
+			_oAuth2ClientManager = oAuth2ClientManager;
+			_rsapiClient = rsapiClient;
+		}
+
+		/// <summary>
+		/// Use this to create an OAuth2 authentication for Relativity, which will allow you to get a Bearer Token and gain access to Relativity
+		/// </summary>
+		/// <param name="username"></param>
+		/// <param name="oAuth2Name"></param>
+		/// <returns></returns>
+		public async Task<Services.Security.Models.OAuth2Client> CreateOAuth2ClientAsync(string username, string oAuth2Name)
+		{
+			DTOs.User user = GetUser(username);
+
+			List<Services.Security.Models.OAuth2Client> oAuth = await _oAuth2ClientManager.ReadAllAsync();
+			if (oAuth.Exists(x => x.Name.Equals(oAuth2Name)))
+			{
+				if (oAuth.Exists(x => x.Name.Equals(oAuth2Name) && x.ContextUser == user.ArtifactID))
+				{
+					Console.WriteLine($"User ({user.EmailAddress}) already has OAuth2 Credentials within ({oAuth2Name})");
+					return oAuth.Find(x => x.Name.Equals(oAuth2Name) && x.ContextUser == user.ArtifactID);
+				}
+
+				throw new Exception($"A different user was has OAuth2 Credentials within ({oAuth2Name})");
+			}
+
+			Services.Security.Models.OAuth2Client oAuthResult = await _oAuth2ClientManager.CreateAsync(oAuth2Name, OAuth2Flow.ClientCredentials, new List<Uri>(), user.ArtifactID);
+			if (oAuthResult != null)
+			{
+				Console.WriteLine($"Created OAuth2 Client Credentials for ({user.EmailAddress}) within ({oAuthResult.Name})");
+				return oAuthResult;
+			}
+			else
+			{
+				throw new Exception($"Failed to create OAuth2 Client Credentials for ({user.EmailAddress}) within ({oAuth2Name})");
+			}
+		}
+
+		/// <summary>
+		/// Use this to get the Bearer Token that will give you access Relativity and various endpoints
+		/// </summary>
+		/// <param name="protocol"></param>
+		/// <param name="serverAddress"></param>
+		/// <param name="clientId"></param>
+		/// <param name="clientSecret"></param>
+		/// <param name="scope"></param>
+		/// <param name="grantType"></param>
+		/// <returns></returns>
+		public async Task<string> GetBearerTokenAsync(string protocol, string serverAddress, string clientId, string clientSecret, string scope = "SystemUserInfo", string grantType = "client_credentials")
+		{

```

Comment:
can we add try/catch and then throw exception with detailed error message for all the methods in the helper. I always try to follow this convention. This will help when there are errors throws in the helper and easily identifiable in the stack trace.

Reply:
Updated please review